src/utp_wrapper.rs

```diff
@@ -32,15 +37,40 @@ impl UtpWrapper {
                 .spawn(move || {
                     let mut socket = socket;
                     socket.set_read_timeout(Some(CHECK_FOR_NEW_WRITES_INTERVAL_MS));
+                    let heartbeat_msg = unwrap_result!(serialise(&CrustMsg::Heartbeat));
+                    let mut heartbeat_deadline = Instant::now() + heartbeat_timeout;
+                    let mut inactivity_deadline = Instant::now() + inactivity_timeout;
+                    let update_heartbeat = |heartbeat_deadline: &mut Instant, inactivity_deadline: &mut Instant| {
+                        let now = Instant::now();
+                        *heartbeat_deadline = now + heartbeat_timeout;
+                        *inactivity_deadline = now + inactivity_timeout;
+                    };
                     'read_write: loop {
                         let mut buf = [0; BUFFER_SIZE];
+                        {
+                            let now = Instant::now();
+                            if now > inactivity_deadline {
+                                info!("Stale connection. Dropping...");
+                                break;
+                            }
+                            if now > heartbeat_deadline {
+                                if let Err(e) = socket.send_to(&heartbeat_msg) {
+                                    error!("Error sending: {:?}", e);
+                                    break;
+                                }
+                                update_heartbeat(&mut heartbeat_deadline,
+                                                 &mut inactivity_deadline);
```

Comment:
`inactivity_deadline` will get updated every time we send a heartbeat. So is it supposed to ever expire?